File: 5_kyu/ISBN10_Validation.go
package kata

func ValidISBN10(isbn string) bool {
	if len(isbn) != 10 {
		// ISBN-10 must be a 10 byte string
		return false
	}
	// calculate checksum
	accu := 0
	for i, c := range isbn[:9] {
		n := int(c - '0')
		if n < 0 || n > 9 {
			return false
		}
		accu += n * (i + 1)
	}
	accu %= 11
	// compare checksum
	check := byte('0' + accu)
	if check > '9' {
		// special case: lower case x is also valid
		return isbn[9] == 'X' || isbn[9] == 'x'
	} else {
		return isbn[9] == check
	}
}

File: 5_kyu/ISBN10_Validation.py
def valid_ISBN10(isbn: str) -> bool: 
    if not (type(isbn) is str and len(isbn) == 10 and isbn[:9].isdigit()):
        return False
    checksum = 0
    for idx in range(9):
        checksum += int(isbn[idx]) * (idx+1)
    checksum %= 11
    checkchar = str(checksum) if checksum < 10 else 'X'
    return isbn[9] == checkchar